substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,763 @@
+			.map_err(|_| Error::<T>::BadState)?;
+
+		T::CoreStaking::update_payee(who, reward_account)?;
+		Self::do_delegate(&proxy_delegator, who, stake.total)
+	}
+
+	fn do_bond(agent_acc: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		let agent = Agent::<T>::get(agent_acc)?;
+
+		let available_to_bond = agent.available_to_bond();
+		defensive_assert!(amount == available_to_bond, "not expected value to bond");
+
+		if agent.is_bonded() {
+			T::CoreStaking::bond_extra(&agent.key, amount)
+		} else {
+			T::CoreStaking::virtual_bond(&agent.key, amount, agent.reward_account())
+		}
+	}
+
+	fn do_delegate(
+		delegator: &T::AccountId,
+		agent: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		let mut ledger = AgentLedger::<T>::get(agent).ok_or(Error::<T>::NotAgent)?;
+		// try to hold the funds.
+		T::Currency::hold(&HoldReason::StakingDelegation.into(), delegator, amount)?;
+
+		let new_delegation_amount =
+			if let Some(existing_delegation) = Delegation::<T>::get(delegator) {
+				ensure!(&existing_delegation.agent == agent, Error::<T>::InvalidDelegation);
+				existing_delegation
+					.amount
+					.checked_add(&amount)
+					.ok_or(ArithmeticError::Overflow)?
+			} else {
+				amount
+			};
+
+		Delegation::<T>::new(agent, new_delegation_amount).update_or_kill(delegator);
+		ledger.total_delegated =
+			ledger.total_delegated.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+		ledger.update(agent);
+
+		Self::deposit_event(Event::<T>::Delegated {
+			agent: agent.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	fn do_release(
+		who: &T::AccountId,
+		delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+		num_slashing_spans: u32,
+	) -> DispatchResult {
+		let mut agent = Agent::<T>::get(who)?;
+		let mut delegation = Delegation::<T>::get(delegator).ok_or(Error::<T>::NotDelegator)?;
+
+		// make sure delegation to be released is sound.
+		ensure!(&delegation.agent == who, Error::<T>::NotAgent);
+		ensure!(delegation.amount >= amount, Error::<T>::NotEnoughFunds);
+
+		// if we do not already have enough funds to be claimed, try withdraw some more.
+		if agent.ledger.unclaimed_withdrawals < amount {
+			// withdraw account.
+			let _ = T::CoreStaking::withdraw_unbonded(who.clone(), num_slashing_spans)
+				.map_err(|_| Error::<T>::WithdrawFailed)?;
+			// reload agent from storage since withdrawal might have changed the state.
+			agent = agent.refresh()?;
+		}
+
+		// if we still do not have enough funds to release, abort.
+		ensure!(agent.ledger.unclaimed_withdrawals >= amount, Error::<T>::NotEnoughFunds);
+
+		// claim withdraw from agent. Kill agent if no delegation left.
+		// TODO(ank4n): Ideally if there is a register, there should be an unregister that should
+		// clean up the agent. Need to improve this.
+		if agent.remove_unclaimed_withdraw(amount)?.update_or_kill()? {
+			let _ = frame_system::Pallet::<T>::dec_providers(who).defensive();
```

Comment:
... and at least defensive assert that the stash was killed there as well.

Reply:
I could check if agent exists as a staker in staking pallet. Though I think we still achieve the same thing.
`dec_providers` would error if there is a consumer reference to it. The consumer reference is cleaned when stash is killed in staking pallet.